### ISSUE TITLE
Using Splash in v1.1.5+ Mobile causes ArgumentException

### ISSUE THREAD

**Issue by bc3tech** [Windows-XAML/Template10#715]
In my Windows 10 Mobile App I've got:

``` c#
        public App()
        {
...
            this.InitializeComponent();

...
            this.SplashFactory = e => new Controls.Splash(e);

            this.UnhandledException += App_UnhandledException;
        }
```

my Controls.Splash is pretty simple:

``` c#
    public sealed partial class Splash : UserControl
    {
        private readonly SplashScreen _splash;

        public Splash(SplashScreen splash)
        {
            _splash = splash;

            this.InitializeComponent();
        }
    }
```

and it actually shows for a spell while loading.
However, quickly after that I'm met with an `ArgumentException` (does not fall within expected range) with this stacktrace:

```
System.ArgumentException occurred
  HResult=-2147024809
  Message=Value does not fall within the expected range.
  InnerException: 
```

at [line 511 of `Bootstrapper.cs`](https://github.com/Windows-XAML/Template10/blob/master/Template10%20%28Library%29/Common/BootStrapper.cs#L511):

``` c#
        public INavigationService NavigationServiceFactory(BackButton backButton, ExistingContent existingContent, Frame frame)
        {
            DebugWrite($"{nameof(backButton)}:{backButton} {nameof(existingContent)}:{existingContent} {nameof(frame)}:{frame}");

            frame.Content = (existingContent == ExistingContent.Include) ? Window.Current.Content : null;
```

Following this back up, we are calling the factory from this block of `InitializeFrameAsync`:

``` c#
            // create the default frame only if there's nothing already there
            // if it is not null, by the way, then the developer injected something & they win
            if (Window.Current.Content == null || Window.Current.Content == splash)
            {
                // build the default frame
                var frame = CreateRootFrame(e);
                var modal = new Controls.ModalDialog
                {
                    DisableBackButtonWhenModal = true,
                    Content = (NavigationServiceFactory(BackButton.Attach, ExistingContent.Include, frame)).FrameFacade.Frame
                };
                Window.Current.Content = modal;
            }
```

_without_ a splash, `frame.Content` is assigned `null`. but **with** one, we try to assign our Splash UserControl in to `frame.Content` and boom.

incidentally I got these details by pulling latest `master` and was able to repro in that state as well.

**Comment by bc3tech** [comment]
I cracked open `Minimal` and attempted to reproduce this issue and **was not able to**. I'm at a loss as to what could be causing this but clearly there's something in my app's flow that's going awry. I'll report back if/when I find it.

**Comment by SergioMorchon** [comment]
It's the same as #675.
Why is this issue closed?

Edit:
If you add:

``` c#
        private bool NeedsNewNavigationService(IActivatedEventArgs args)
        {
            switch (args.PreviousExecutionState)
            {
                case ApplicationExecutionState.ClosedByUser:
                case ApplicationExecutionState.NotRunning:
                case ApplicationExecutionState.Terminated:
                    return true;
                case ApplicationExecutionState.Running:
                case ApplicationExecutionState.Suspended:
                    return false;
            }
            return true;
        }
```

and check for it on the `OnInitializeAsync` App method _before_ create a new Window content, it will run ok :)

In my case:

``` c#
        private bool NeedsNewNavigationService(IActivatedEventArgs args)
        {
            switch (args.PreviousExecutionState)
            {
                case ApplicationExecutionState.ClosedByUser:
                case ApplicationExecutionState.NotRunning:
                case ApplicationExecutionState.Terminated:
                    return true;
                case ApplicationExecutionState.Running:
                case ApplicationExecutionState.Suspended:
                    return false;
            }
            return true;
        }

        public override async Task OnInitializeAsync(IActivatedEventArgs args)
        {
            InstallVoiceCommands();
            if (NeedsNewNavigationService(args))
            {
                var nav = NavigationServiceFactory(BackButton.Attach, ExistingContent.Include);

                if (Window.Current.Content is ModalDialog)
                {
                    Window.Current.Content = new ModalDialog
                    {
                        DisableBackButtonWhenModal = true,
                        Content = new Views.Shell(nav),
                        ModalContent = new Views.Busy()
                    };
                }
                else
                {
                    Window.Current.Content = new Views.Shell(nav);
                }
            }
            await Task.CompletedTask;
        }
```

**Comment by bc3tech** [comment]
good info @SergioMorchon. I closed this issue because I couldn't reproduce it with the Minimal template which seems to point the finger at something in **my** app code vs something in Template 10.

Incidentally, adding code in my app that (it would seem) effectively does the same thing had _no_ issues at all!

```
        async public override Task OnInitializeAsync(IActivatedEventArgs args)
        {
            Window.Current.Content = new Controls.Splash(args.SplashScreen);
            Window.Current.Activate();
...
```

then

```
        public override async Task OnStartAsync(StartKind startKind, IActivatedEventArgs args)
        {
...
            var nav = NavigationServiceFactory(BackButton.Attach, ExistingContent.Exclude);
            nav.Navigate(typeof(MainPage));
            Window.Current.Content = nav.FrameFacade.Frame;
            Window.Current.Activate();
        }
```

Try as I might, I couldn't pinpoint exactly what was causing the problem. Do you see the blowup occur on the assignment of `Frame.Content` down in T10's Bootstrapper as well?

cc @JerryNixon 

Additionally, I see your fix takes care of _two other_ issues reported on the library. I'm not sure if the issues are still outstanding in the latest nuget pkg (v1.1.8) but I think we should take a hard look at pulling in your changes. I'm going to make the changes locally and see if they do indeed take care of my problem and report back. If you could, you should fork the repo, create a branch, put your change in, and submit a PR mentioning all 3 of the issues in which you've reported this fix.
Thanks!

**Comment by SergioMorchon** [comment]
I figured out that this exception was caused when you try to add an element with the same name.
You can read about it [here](https://digitaltoolfactory.net/blog/2009/10/how-to-fix-value-does-not-fall-within-the-expected-range-error-in-silverlight/).
The NavigationServiceFactory might be causing this bug if the application is already running (or from Suspended). That means, the whole App was in memory, not terminated, so it is unnecesary to regenerate the navigation service nor any "structural" code.
If the app comes from Suspended (and not terminated), the developer probably:
- Wants to do nothing.
- Wants to update time-susceptible data.

In conclusion:
I wouldn't treat this as a bug. BUT it is very weird that a NavigationServiceFactory method throws this choerence-less exception.

**Comment by bc3tech** [comment]
@SergioMorchon alas, this did not fix my issue. Here's how I rolled it in to T10:
We initialize the frame (set window content) in `InitializeFrameAsync` so I did this:

```
        private async Task InitializeFrameAsync(IActivatedEventArgs e)
        {
            DebugWrite($"{nameof(IActivatedEventArgs)}:{e.Kind}");

            // first show the splash 
            FrameworkElement splash = null;
            if (SplashFactory != null && e.PreviousExecutionState != ApplicationExecutionState.Suspended
                && NeedsNewNavigationService(e))
            {
                Window.Current.Content = splash = SplashFactory(e.SplashScreen);
                Window.Current.Activate();
            }
```

Note the new call to `NeedsNewNavigationService` - in this method I have your code:

```
        private bool NeedsNewNavigationService(IActivatedEventArgs args)
        {
            switch (args.PreviousExecutionState)
            {
                case ApplicationExecutionState.ClosedByUser:
                case ApplicationExecutionState.NotRunning:
                case ApplicationExecutionState.Terminated:
                    return true;
                case ApplicationExecutionState.Running:
                case ApplicationExecutionState.Suspended:
                    return false;
            }
            return true;
        }
```

Then, in my consuming App, I did the `SplashFactory` usage per T10 pattern:

```
        public App()
        {
            this.InitializeComponent();

            this.UnhandledException += App_UnhandledException;
...
            this.SplashFactory = e => new Controls.Splash(e);
        }
```

```
        public override async Task OnStartAsync(StartKind startKind, IActivatedEventArgs args)
        {
...
            this.NavigationService.Navigate(typeof(MainPage));
        }
```

But I'm still met with this same exception.

**Comment by callummoffat** [comment]
I'm getting this issue on Windows 10 Mobile, while using Azure Mobile Apps authentication. Whenever authentication is triggered, it updates Window.Current.Content to a login UserControl, which causes the application to crash. None of the currently suggested solutions have any effect, and for my app to work, I **need** authentication.

**Comment by bc3tech** [comment]
@callummoffat here is what i'm doing right now, in case it can help directly or indirectly:

```
        async public override Task OnInitializeAsync(IActivatedEventArgs args)
        {
            if (args.PreviousExecutionState != ApplicationExecutionState.Running
                && args.PreviousExecutionState != ApplicationExecutionState.Suspended)
            {
                Window.Current.Content = new Controls.Splash(args.SplashScreen); 
                Window.Current.Activate();
            }
...
        }
```

then

```
        public override Task OnStartAsync(StartKind startKind, IActivatedEventArgs args)
        {
            if (Window.Current.Content is Controls.Splash)
            {
                var nav = NavigationServiceFactory(BackButton.Attach, ExistingContent.Exclude);
                nav.Navigate(typeof(MainPage));
                Window.Current.Content = nav.FrameFacade.Frame;
                Window.Current.Activate();
            }

            return Task.CompletedTask;
        }
```

**Comment by callummoffat** [comment]
I'll test this, and see if it helps. I'm not sure if the fix can be integrated into my app, since the method that changes Window.Current.Content is external in this case (MobileServiceClient.LoginAsync()); I think this issue needs to be fixed in Template 10 itself.

**Comment by bc3tech** [comment]
agreed. just thought it might help shed some light _if_ you were able to do anything.

**Comment by callummoffat** [comment]
Nope, the fix didn't seem to do anything. For this to work, I need the issue to be fixed in Template10. 

@JerryNixon, just to recap:

 This issue causes an ArgumentException when Window.Current.Content is set multiple times on a Windows 10 Mobile device. This can be worked around in some cases, but if the offending method is contained within external code (ie: in a Nuget library), there doesn't appear to be a workaround. The cases where a workaround is impossible include authentication using the Azure Mobile Apps framework.

@Windows-XAML/community, do you have any ideas has to how to fix this?
